AppLovin MAX Demo App - Swift/AppLovin MAX Demo App - Swift/MAX/Native Ads/UI/ALMAXSwiftUITemplateNativeAdView.swift

```diff
@@ -0,0 +1,145 @@
+                                                 didFailToLoadNativeAd: viewModel.didFailToLoadNativeAd(forAdUnitIdentifier:withError:),
+                                                 didClickNativeAd: viewModel.didClickNativeAd(_:),
+                                                 didExpireNativeAd: viewModel.didExpireNativeAd(_:),
+                                                 didPayRevenue: viewModel.didPayRevenue(for:))
+                .frame(width: 300, height: 250)
+            
+            callbacksTable
+                .frame(maxHeight: .infinity)
+            Button {
+                viewModel.triggerShowAd = true
+            } label: {
+                Text("Show")
+            }
+        }
+    }
+    
+    var callbacksTable: some View {
+        List(viewModel.callbacks) {
+            Text($0.callback)
+        }
+    }
+}
+
+@available(iOS 13.0, *)
+class ALMAXSwiftUITemplateNativeAdViewModel: NSObject, ObservableObject
+{
+    @Published fileprivate var callbacks: [CallbackTableItem] = []
+    @Published var triggerShowAd: Bool = false
+    
+    let adUnitIdentifier: String
+    
+    let adLoader: NativeSwiftUIAdLoader
+    
+    var containerView = UIView()
+    
+    init(adUnitId: String)
+    {
+        self.adUnitIdentifier = adUnitId
+        self.adLoader = NativeSwiftUIAdLoader(adUnitId: adUnitId, containerView: containerView)
+    }
+    
+    private func logCallback(functionName: String = #function)
+    {
+        DispatchQueue.main.async {
+            withAnimation {
+                self.callbacks.append(CallbackTableItem(callback: functionName))
+            }
+        }
+    }
+}
+
+
+
+
+@available(iOS 13.0, *)
+extension ALMAXSwiftUITemplateNativeAdViewModel: MANativeAdDelegate
+{
+    // MARK: MANativeAdDelegate Protocol
+
+    func didLoadNativeAd(_ nativeAdView: MANativeAdView?, for ad: MAAd)
+    {
+        logCallback()
+        
+        // Save ad for clean up
+        self.adLoader.nativeAd = ad
+        
+        if let adView = nativeAdView
+        {
+            // Add ad view to container view.
+            self.adLoader.replaceCurrentNativeAdView(adView)
+            self.containerView.addSubview(adView)
+            
+            // Set to false after adding the ad view to your layout if modifying constraints.
+            adView.translatesAutoresizingMaskIntoConstraints = false
+            
+            // Set ad view to span width and height of container and center the ad
+            self.containerView.widthAnchor.constraint(equalTo: adView.widthAnchor).isActive = true
+            self.containerView.heightAnchor.constraint(equalTo: adView.heightAnchor).isActive = true
+            self.containerView.centerXAnchor.constraint(equalTo: adView.centerXAnchor).isActive = true
+            self.containerView.centerYAnchor.constraint(equalTo: adView.centerYAnchor).isActive = true
+            
+            self.triggerShowAd = false
```

Comment:
We can remove the `self` prefix for all these instance properties since it's optional in Swift. Good rule of thumb is omit the prefix unless there's ambiguity that the compiler complains about (conflict with local variable name)